packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2057,6 +2082,48 @@ export class KeyringController extends BaseController<
     this.messagingSystem.publish(`${name}:unlock`);
   }
 
+  /**
+   * Execute the given function after acquiring the controller lock
+   * and save the keyrings to state after it, or rollback to their
+   * previous state in case of error.
+   *
+   * @param fn - The function to execute.
+   * @returns The result of the function.
+   */
+  async #persistOrRollback<T>(fn: MutuallyExclusiveCallback<T>): Promise<T> {
+    return this.#withRollback(async ({ releaseLock }) => {
+      const callbackResult = await fn({ releaseLock });
+      // State is committed only if the operation is successful
+      await this.#updateVault();
+
+      return callbackResult;
+    });
+  }
+
+  /**
+   * Execute the given function after acquiring the controller lock
+   * and rollback keyrings and password states in case of error.
+   *
+   * @param fn - The function to execute atomically.
+   * @returns The result of the function.
+   */
+  async #withRollback<T>(fn: MutuallyExclusiveCallback<T>): Promise<T> {
+    return this.#withControllerLock(async ({ releaseLock }) => {
+      const currentSerializedKeyrings = await this.#getSerializedKeyrings();
+      const currentPassword = this.#password;
+
+      try {
+        return await fn({ releaseLock });
+      } catch (e) {
+        // Keyrings and password are restored to their previous state
+        await this.#restoreSerializedKeyrings(currentSerializedKeyrings);
```

Comment:
I'm realising now that this probably does not restore the QR keyring event listener, because that's a specific step of `submitPassword` and `submitEncryptionKey` 🤔

We need to move the subscription to the internal restore method

Reply:
Good catch

Reply:
Probably it makes sense for that to go in the internal restore method

Reply:
Moving it in `#newKeyring` allowed me to also remove some superfluous code, since the subscription was also created in `#addQRKeyring`: [36347ad](https://github.com/MetaMask/core/pull/4192/commits/36347ad5dceee4a8477c699e20093137d03fb680)